pandas-stubs/core/groupby/grouper.pyi

```diff
@@ -7,67 +14,115 @@ from pandas import (
     Series,
 )
 from pandas.core.groupby.ops import BaseGrouper
+from pandas.core.resample import TimeGrouper
+from typing_extensions import (
+    Self,
+    deprecated,
+)
 
-from pandas._typing import NDFrameT
+from pandas._libs.lib import NoDefault
+from pandas._typing import (
+    ArrayLike,
+    Axis,
+    Frequency,
+    Incomplete,
+    KeysArgType,
+    Level,
+    ListLikeHashable,
+    NDFrameT,
+    npt,
+)
+from pandas.util._decorators import cache_readonly
 
 class Grouper:
-    def __new__(cls, *args, **kwargs): ...
-    key = ...
-    level = ...
-    freq = ...
-    axis = ...
-    sort = ...
-    grouper = ...
-    obj = ...
-    indexer = ...
-    binner = ...
+    key: KeysArgType | None
+    level: Level | ListLikeHashable[Level] | None
+    freq: Frequency | None
+    axis: Axis
+    sort: bool
+    dropna: bool
+    binner: Incomplete
+    @overload
+    def __new__(
+        cls,
+        key: KeysArgType | None = None,
+        level: Level | ListLikeHashable[Level] | None = None,
+        axis: Axis | NoDefault = ...,
+        sort: bool = False,
+        dropna: bool = True,
+    ) -> Self: ...
+    @overload
+    def __new__(cls, *args, freq: Frequency, **kwargs) -> TimeGrouper: ...
     def __init__(
```

Comment:
I suggest keeping `__init__` for now. `__new__` that returns different types from `__new__` than the class itself is known to be a source of compatibility issues between type checkers. Until this behavior is standardized, we can keep `__init__` because all type checkers are happy with it.

Reply:
We use that "feature" of different `__new__()` methods to allow different types to be returned in many places in `pandas-stubs`.  So since you know that the argument `freq` being present changes what type of object is returned, we should take advantage of that.  I'd prefer just having `__new__()` in this case.